### PR TITLE
add in-circuit EdDSA verification gadget

### PR DESCRIPTION
**Pull request by arnaucube**
- add in-circuit EdDSA verification gadget compatible with the already existing out-circuit logic (in-circuit gadget offered now under the `r1cs` feature)
- modifies the sign & verify interfaces to instead of getting as input `message: &[E]`, they now get `message: &TE::BaseField`, to be closer to other implementations (like circomlib & compatible ones)
- moves the `poseidon_config` function outside from tests, since when externally using this lib usually the user will also use the `poseidon_config` helper

### CODE REVIEW

File: src/constraints.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    cs: ConstraintSystemRef<CF<C>>,
+    poseidon_config: PoseidonConfig<CF<C>>,
+    pk: GC,
+    sig: (GC, NonNativeFieldVar<C::ScalarField, CF<C>>),
+    msg: FpVar<CF<C>>,
+) -> Result<Boolean<CF<C>>, Error>
+where
+    C: CurveGroup,
+    GC: CurveVar<C, CF<C>> + ToConstraintFieldGadget<CF<C>>,
+{
```
Comment (kilic) [inline src/constraints.rs:28]: Can we return `ark_relations::r1cs::Result` there and get rid of unwraps?

File: src/signature.rs [hunk 2]
```diff
@@ -6,8 +6,8 @@ use ark_serialize::CanonicalSerialize;
 /// `SignatureComponents` contains the realized parts of a signature
 #[derive(Copy, Clone, Debug, CanonicalSerialize, CanonicalDeserialize)]
 pub struct Signature<TE: TECurveConfig + Clone> {
-    r: Affine<TE>,
-    s: TE::ScalarField,
+    pub r: Affine<TE>,
```
Comment (kilic) [inline src/signature.rs]: There are getter methods but I can give pass anyway

File: src/eddsa.rs [hunk 3]
```diff
@@ -45,7 +45,7 @@ impl SecretKey {
 
 /// `PublicKey` is EdDSA signature verification key
 #[derive(Copy, Clone, Debug, CanonicalSerialize, CanonicalDeserialize)]
-pub struct PublicKey<TE: TECurveConfig>(Affine<TE>);
+pub struct PublicKey<TE: TECurveConfig>(pub Affine<TE>);
```
Comment (kilic) [inline src/eddsa.rs]: Could use existing `as_ref` method to reach inner

**Comment by kilic** [comment]
Some small change request, overall it's lgtm

File: src/constraints.rs (hunk 1 above)
Comment (arnaucube) [reply src/constraints.rs:28]: Right! my bad. Done at https://github.com/kilic/arkeddsa/pull/1/commits/15158474ab0eea56f40aac9a316444b03ff6d4eb

File: src/signature.rs (hunk 2 above)
Comment (arnaucube) [reply src/signature.rs]: cool, returned to private param and using `.r()` method at https://github.com/kilic/arkeddsa/pull/1/commits/15158474ab0eea56f40aac9a316444b03ff6d4eb

File: src/eddsa.rs (hunk 3 above)
Comment (arnaucube) [reply src/eddsa.rs]: cool, removed the `pub` and used the `as_ref` method at https://github.com/kilic/arkeddsa/pull/1/commits/15158474ab0eea56f40aac9a316444b03ff6d4eb

**Pull request merged**